src/concurrency.md

```diff
@@ -232,6 +233,96 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[threadpool-walk]: #threadpool-walk
+<a name="threadpool-walk"></a>
+## Use a threadpool to calculate SHA1 to all *.iso files in a folder
+
+[![threadpool-badge]][threadpool] [![num_cpus-badge]][num_cpus] [![walkdir-badge]][walkdir] [![cat-concurrency-badge]][cat-concurrency][![cat-filesystem-badge]][cat-filesystem]
+
+This example uses `threadpool`, `walkdir` and `ring` crates to calculate the SHA1 for every file present in the current directory. A threadpool `pool` is created using the number of cpus present in the system. Then every `entry` returned by `walkdir` is passed into this pool to perform the operations of reading and compute SHA1. At the end the program waits for all jobs to finish. To get better results, compile this program in release mode.
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate walkdir;
+extern crate ring;
+extern crate num_cpus;
+extern crate threadpool;
+
+# error_chain! {
+#     foreign_links {
+#         Io(std::io::Error);
+#     }
+# }
+
+use walkdir::{WalkDir, DirEntry};
+use std::fs::File;
+use std::io::BufReader;
+use std::io::prelude::*;
+use threadpool::ThreadPool;
+use std::sync::mpsc::channel;
+use ring::digest;
+
+# // Verify the iso extension
+# fn is_iso(entry: &DirEntry) -> bool {
+#     entry.file_name()
+#         .to_str()
+#         .map(|s| s.ends_with(".iso"))
+#         .unwrap_or(false)
+# }
+
+fn run() -> Result<()> {
+    let cpus = num_cpus::get();
+    let pool = ThreadPool::new(cpus);
+    let mut file_counter: usize = 0;
+
+    let (tx, rx) = channel();
+
+    // Look in the current directory.
+    for entry in WalkDir::new(".")
+        .follow_links(true)
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| !e.file_type().is_dir() && is_iso(e)) {
+            // Need to know how many entry were sent to the pool.
+            file_counter = file_counter + 1;
+            let entry = entry.clone();
+            let tx = tx.clone();
+            pool.execute(move || {
+                let file = match File::open(entry.path()) {
+                    Ok(file) => file,
+                    Err(err) => {
+                        tx.send(Err(err)).expect("Could not send data!");
+                        return;
+                    },
+                };
+                let mut buf_reader = BufReader::new(file);
```

Comment:
Reading the whole file into memory is not really the best way to go. Ring provides a nice incremental api.
for its usage please see the 
[calculate the sha-256 digest of a file](https://rust-lang-nursery.github.io/rust-cookbook/basics.html#calculate-the-sha-256-digest-of-a-file) example

Moreover it would be much nicer to extract the file reading and sha1 calculation into a separate function taking a `AsRef<Path>` and returning a `Result`. Then you would have only one `tx.send()` without the verbose matches all over the place.